### PR TITLE
[Serializer] Argument objects

### PR DESCRIPTION
**Pull request by theofidry**
| Q | A |
| --- | --- |
| Branch? | 3.1 |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | TODO |
| Fixed tickets | none |
| License | MIT |
| Doc PR | TODO |

Assuming with have the two following entities:

``` php
namespace AppBundle\Entity;

class Dummy
{
    public function __construct(int $id, string $name, string $email, AnotherDummy $anotherDummy)
    {
        $this->id = $id;
        $this->name = $name;
        $this->email = $email;
        $this->anotherDummy = $anotherDummy;
    }
}

class AnotherDummy
{
    public function __construct(int $id, string $uuid, bool $isEnabled)
    {
        $this->id = $id;
        $this->uuid = $uuid;
        $this->isEnabled = $isEnabled;
    }
}
```

Doing the following will fail:

``` php
$serializer->denormalize(
    [
        'id' => $i,
        'name' => 'dummy',
        'email' => 'du@ex.com',
        'another_dummy' => [
            'id' => 1000 + $i,
            'uuid' => 'azerty',
            'is_enabled' => true,
        ],
    ],
    \AppBundle\Entity\Dummy::class
);
```

with a type error, because the 4th argument passed to `Dummy::__construct()` will be an array. The following patch checks if the type of the argument is an object, and if it is tries to denormalize that object as well.

I'm not sure if it's me missing something or this is a use case that has been omitted (willingly or not), but if it's a valuable patch I would be happy to work on finishing it.

### CODE REVIEW

File: src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php [hunk 1]
```diff
@@ -319,9 +319,13 @@ protected function instantiateObject(array &$data, $class, array &$context, \Ref
                         $params = array_merge($params, $data[$paramName]);
                     }
                 } elseif ($allowed && !$ignored && (isset($data[$key]) || array_key_exists($key, $data))) {
-                    $params[] = $data[$key];
-                    // don't run set for a parameter passed to the constructor
-                    unset($data[$key]);
```
Comment (dunglas) [inline src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php]: Why removing the unset?

Comment (theofidry) [inline src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php]: bad copy/paste, should be kept

File: src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php [hunk 2]
```diff
@@ -319,9 +319,13 @@ protected function instantiateObject(array &$data, $class, array &$context, \Ref
...
+                    $parameterData = $data[$key];
+                    if (null !== $constructogrParameter->getClass()) {
+                        $parameterData = $this->serializer->denormalize($parameterData, $constructorParameter->getClass()->getName(), null, $context);
```
Comment (dunglas) [inline src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php]: I would add `$format` as a new optional paremeter for `instantiateObject` and pass it to the serializer.

**Comment by dunglas** [comment]
This case was omitted when adding deserialization support for complex relations (in 3.1). IMO it should be merged in 3.1.

Can you add a test for this?

File: src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php (hunk 2 above)
Comment (dunglas) [inline src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php]: Forgot this comment, it would be a BC break: `Add argument with a default value	No`

File: src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php [hunk 3]
```diff
@@ -281,12 +281,13 @@ protected function getConstructor(array &$data, $class, array &$context, \Reflec
...
      * @param \ReflectionClass $reflectionClass
      * @param array|bool       $allowedAttributes
+     * @param string|null      $format
      *
      * @return object
      *
      * @throws RuntimeException
      */
-    protected function instantiateObject(array &$data, $class, array &$context, \ReflectionClass $reflectionClass, $allowedAttributes)
+    protected function instantiateObject(array &$data, $class, array &$context, \ReflectionClass $reflectionClass, $allowedAttributes, $format = null)
```
Comment (dunglas) [inline src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php]: Sorry I realized that this is BC break... It should be removed. But it's annoying because some normalizers may depend of the format (normalizers of API Platform for instance).

An alternative approach: 

* Rename this method `instantiateComplexObject` (with the new parameter, it can be added before `$context`)
* Add back `instantiateObject` with the current signature and call `instantiateComplexObject` inside it with format as `null`
* Deprecate `instantiateObject`

WDYT?

File: src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php [hunk 4]
```diff
@@ -319,8 +320,13 @@ protected function instantiateObject(array &$data, $class, array &$context, \Ref
                         $params = array_merge($params, $data[$paramName]);
                     }
                 } elseif ($allowed && !$ignored && (isset($data[$key]) || array_key_exists($key, $data))) {
-                    $params[] = $data[$key];
-                    // don't run set for a parameter passed to the constructor
+                    $parameterData = $data[$key];
+                    if (null !== $constructorParameter->getClass()) {
+                        $parameterData = $this->serializer->denormalize($parameterData, $constructorParameter->getClass()->getName(), null, $context);
```
Comment (theofidry) [inline src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php]: @dunglas I'm doing a `denormalize` but the normalizer is... `SerializerAware`, not `NormalizerAware`. This works because of the way the serializer is setup but it's violating the interfaces. What should we do about it? Calling `deserializer` even though it will be more expensive?

Comment (theofidry) [inline src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php]: Also calling `deserialize()` will fail if you are using a serializer without any encoder, which might be the case when you are dealing with arrays and php objects, i.e. are not taking any JSON/XML input. Not likely in a regular app, but can happen in libraries.

Comment (dunglas) [inline src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php]: Maybe can you add a `instanceof NormalizerInterface` check?

Comment (theofidry) [inline src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php]: Cant' we make the normalizer `NormalizerAware` instead and deprecate it's `SerializerAware`?

File: src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php (hunk 3 above)
Comment (theofidry) [inline src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php]: will do

**Comment by theofidry** [comment]
Hm, looks like this also occurs for setters

File: src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php (hunk 4 above)
Comment (dunglas) [inline src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php]: It would be better. But it should be done in another PR.

File: src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php [hunk 5]
```diff
@@ -281,12 +290,13 @@ protected function getConstructor(array &$data, $class, array &$context, \Reflec
...
      * @param \ReflectionClass $reflectionClass
      * @param array|bool       $allowedAttributes
+     * @param string|null      $format
      *
      * @return object
      *
      * @throws RuntimeException
      */
-    protected function instantiateObject(array &$data, $class, array &$context, \ReflectionClass $reflectionClass, $allowedAttributes)
+    protected function instantiateComplexObject(array &$data, $class, array &$context, \ReflectionClass $reflectionClass, $allowedAttributes, $format = null)
```
Comment (dunglas) [inline src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php]: It can now be:
```php
instantiateComplexObject(array &$data, $class, $format, array &$context, \ReflectionClass $reflectionClass, $allowedAttributes)
```

To mimic the `SerializerInterface::deserialize` prototype.

File: src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php [hunk 6]
```diff
@@ -318,28 +328,43 @@ protected function instantiateObject(array &$data, $class, array &$context, \Ref
...
-                    $params[] = $data[$key];
-                    // don't run set for a parameter passed to the constructor
+
+                    continue;
+                }
+
+                if ($allowed && !$ignored && (isset($data[$key]) || array_key_exists($key, $data))) {
+                    $parameterData = $data[$key];
+                    if (null !== $constructorParameter->getClass()) {
+                        $parameterData = $this->serializer->deserialize($parameterData, $constructorParameter->getClass()->getName(), null, $context);
```
Comment (dunglas) [inline src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php]: `$format` instead of `null`

File: src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php (hunk 5 above)
Comment (theofidry) [inline src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php]: my bad I kept the old signature, will change it

**Comment by theofidry** [comment]
@dunglas not sure where the problem lies right now, will have to dig a bit into it later although I hope there won't be any big change.

**Comment by theofidry** [comment]
@dunglas tests are passing

File: src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php [hunk 7]
```diff
@@ -269,6 +269,15 @@ protected function getConstructor(array &$data, $class, array &$context, \Reflec
     }
 
     /**
+     * @see instantiateComplexObject
+     * @deprecated Since 3.1, will be removed in 4.0. Use instantiateComplexObject instead.
```
Comment (xabbuh) [inline src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php]: missing `@trigger_error()` call for this

Comment (theofidry) [inline src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php]: @xabbuh done

File: src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php (hunk 5 above)
Comment (Ener-Getick) [inline src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php]: why not using a bc layer to keep the old name ``instantiateObject`` ? it looks clearer imo

File: src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php [hunk 8]
```diff
@@ -318,28 +330,43 @@ protected function instantiateObject(array &$data, $class, array &$context, \Ref
...
-                } elseif ($allowed && !$ignored && (isset($data[$key]) || array_key_exists($key, $data))) {
-                    $params[] = $data[$key];
-                    // don't run set for a parameter passed to the constructor
+
+                    continue;
+                }
+
+                if ($allowed && !$ignored && (isset($data[$key]) || array_key_exists($key, $data))) {
+                    $parameterData = $data[$key];
+                    if (null !== $constructorParameter->getClass()) {
```
Comment (Ener-Getick) [inline src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php]: ``ReflectionParameter::getType()`` should be used when available imo.

File: src/Symfony/Component/Serializer/composer.json [hunk 9]
```diff
@@ -16,12 +16,12 @@
         }
     ],
     "require": {
-        "php": ">=5.5.9"
+        "php": ">=5.5.9",
+        "symfony/property-access": "^3.1"
```
Comment (Ener-Getick) [inline src/Symfony/Component/Serializer/composer.json]: not sure this should be moved as the serializer can be used without it with some normalizers.

Comment (theofidry) [inline src/Symfony/Component/Serializer/composer.json]: Ah my bad, this one should not have been committed

File: src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php [hunk 10]
```diff
@@ -269,6 +269,17 @@ protected function getConstructor(array &$data, $class, array &$context, \Reflec
     }
 
     /**
+     * @see instantiateComplexObject
+     * @deprecated Since 3.1, will be removed in 4.0. Use instantiateComplexObject instead.
+     */
+    protected function instantiateObject(array &$data, $class, array &$context, \ReflectionClass $reflectionClass, $allowedAttributes)
+    {
+        @trigger_error(sprintf('"%s()" has been deprecated sin Symfony 3.1 and will be removed in version 4.0. Use "%s::instantiateComplexObject()" instead.', __METHOD__, __CLASS__));
```
Comment (theofidry) [inline src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php]: I'm quite surprised this went undetected by the phpunit bridge...

Comment (WouterJ) [inline src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php]: @theofidry phpunit bridge only discovers deprecated errors, not general silenced errors.

File: src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php (hunk 8 above)
Comment (theofidry) [inline src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php]: I personally find it easier to check check if the class is null or not in this case

File: src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php (hunk 10 above)
Comment (theofidry) [inline src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php]: @WouterJ but I have a test with the `@legacy` group, so I guess I expected it to warn me or throw me an error if no deprecation notice were discovered

**Comment by theofidry** [comment]
Updated

File: src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php (hunk 8 above)
Comment (Ener-Getick) [inline src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php]: I agree but the type hint may be on a class not loadable. In this case, the exception thrown should be catched and the default value applied.

Comment (theofidry) [inline src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php]: could you elaborate? On which scenario may it happen? Not that I'm not trusting you, but if that's the case I would like to add a test for it :)

Comment (Ener-Getick) [inline src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php]: In case you use a type hint with a class from a third party library which is not loaded because you don't need it for example.
And the advantage of getType() is that it doesn't need to load the class, so would work anyway.

File: src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php [hunk 11]
```diff
@@ -318,28 +330,43 @@ protected function instantiateObject(array &$data, $class, array &$context, \Ref
...
+                        'Cannot create an instance of %s from serialized data because its constructor requires parameter "%s" to be present.',
+                        $class,
+                        $constructorParameter->name
+                    )
+                );
             }
 
             if ($constructor->isConstructor()) {
                 return $reflectionClass->newInstanceArgs($params);
-            } else {
```
Comment (Ener-Getick) [inline src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php]: Should not be updated imo as this less clear...

File: src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php (hunk 8 above)
Comment (theofidry) [inline src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php]: Ah I see, fair enough will change that then

File: src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php (hunk 11 above)
Comment (theofidry) [inline src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php]: Personally the more indent you have the harder it is to read. And if the choice is to use `if/else/elseif` blocks, then everything should be in an else instead of the return done L311.

Comment (Ener-Getick) [inline src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php]: There is only one line here and i agree that most of the time ``else`` +``return`` is only redundant but here it highlights the fact that one of this line will be executed and that the second one is not just a default.

File: src/Symfony/Component/Serializer/composer.json [hunk 12]
```diff
@@ -21,7 +21,6 @@
     "require-dev": {
         "symfony/yaml": "~2.8|~3.0",
         "symfony/config": "~2.8|~3.0",
-        "symfony/property-access": "~2.8|~3.0",
```
Comment (Ener-Getick) [inline src/Symfony/Component/Serializer/composer.json]: Should not be removed i guess ?

File: src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php (hunk 11 above)
Comment (theofidry) [inline src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php]: >Edit: BTW, we could reverse the condition if ($constructor) to have less lines indented as you said :)

I like this one :)

File: src/Symfony/Component/Serializer/Normalizer/GetSetMethodNormalizer.php [hunk 13]
```diff
@@ -47,7 +47,7 @@ public function denormalize($data, $class, $format = null, array $context = arra
         $normalizedData = $this->prepareForDenormalization($data);
 
         $reflectionClass = new \ReflectionClass($class);
-        $object = $this->instantiateObject($normalizedData, $class, $context, $reflectionClass, $allowedAttributes);
+        $object = $this->instantiateComplexObject($normalizedData, $class, $context, $reflectionClass, $allowedAttributes, $format);
```
Comment (Ener-Getick) [inline src/Symfony/Component/Serializer/Normalizer/GetSetMethodNormalizer.php]: Unfortunatelly this looks like a bc break as people may depend on the call to ``instantiateObject``.

File: src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php (hunk 11 above)
Comment (Ener-Getick) [inline src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php]: But not sure this would be accepted as this is a small detail and could create many merge conflicts.

File: src/Symfony/Component/Serializer/Normalizer/GetSetMethodNormalizer.php (hunk 13 above)
Comment (theofidry) [inline src/Symfony/Component/Serializer/Normalizer/GetSetMethodNormalizer.php]: Not sure to see what you mean, before it was simply failing so people could not used it. The method has been changed because there's a BC break in the method signature (we are adding `$format`), but the behaviour is the previous one.

Comment (Ener-Getick) [inline src/Symfony/Component/Serializer/Normalizer/GetSetMethodNormalizer.php]: In case someone extended ``instantiateObject``, it won't be called anymore. I'll provide you an example of how this could be managed.

File: src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php (hunk 11 above)
Comment (theofidry) [inline src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php]: Yeah actually I think I'll revert everything for that part, it's just CS and although I think it makes a big difference it terms of readability, I fear that it will cause lots of conflicts later on. One solution would be do go back in earlier versions and fix that version after version, but it's rather tedious and slow.

File: src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php (hunk 5 above)
Comment (Ener-Getick) [inline src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php]: What do you think about using ``instantiateClass`` to have something shorter but still clear ?

File: src/Symfony/Component/Serializer/Normalizer/GetSetMethodNormalizer.php (hunk 13 above)
Comment (Ener-Getick) [inline src/Symfony/Component/Serializer/Normalizer/GetSetMethodNormalizer.php]: You could do something like:
```php
protected function instantiateObject(..., $allowedAttributes/**, $format*/)
{
    $format = null;
    if (func_num_args() >= 6) {
        $format = func_get_arg(5);
    }

    // ...
}

protected function instantiateComplexObject(...)
{
    static $deprecationTriggered = false;
    if (false === $deprecationTriggered) {
        $deprecationTriggered = true;
        $class = get_class($this);
        if (0 !== strpos($class, 'Symfony\\') 
            && self::class !== (new ReflectionMethod($this, 'instantiateObject'))->getDeclaringClass()->getName()) {
            @trigger_error(..., E_USER_DEPRECATED);
        }
    }

    return $this->instanciateObject(...);
}

Comment (theofidry) [inline src/Symfony/Component/Serializer/Normalizer/GetSetMethodNormalizer.php]: taking `$format` as the 6th argument in `instantiateObject` would still be a problem if you extended it

Comment (theofidry) [inline src/Symfony/Component/Serializer/Normalizer/GetSetMethodNormalizer.php]: @dunglas what should we do about this one? Sounds like the easiest way would be to have a new `GetSetNormalizer` and deprecate the old one but 1. this is super ugly and 2. it means finding a new name... which sucks.

Comment (theofidry) [inline src/Symfony/Component/Serializer/Normalizer/GetSetMethodNormalizer.php]: and we actually have the same problem for `AbstractObjectNormalizer`...

Comment (Ener-Getick) [inline src/Symfony/Component/Serializer/Normalizer/GetSetMethodNormalizer.php]: Why would it be a problem ?

Comment (dunglas) [inline src/Symfony/Component/Serializer/Normalizer/GetSetMethodNormalizer.php]: Changing the class name looks like a no go for me...

@Ener-Getick changing the signature of a `propected` method is a BC break isn't it? (not sure of how PHP handle the inheritance of the splat operator).

Comment (dunglas) [inline src/Symfony/Component/Serializer/Normalizer/GetSetMethodNormalizer.php]: The only solution that comes to my mind is to set the `$format` to null for now and fix that behavior in Symfony 4... Maybe someone in @symfony/deciders has a better idea.

Comment (Ener-Getick) [inline src/Symfony/Component/Serializer/Normalizer/GetSetMethodNormalizer.php]: @dunglas it is but we can use ``func_num_args`` and ``func_get_arg`` to avoid changing its signature.

Comment (dunglas) [inline src/Symfony/Component/Serializer/Normalizer/GetSetMethodNormalizer.php]: Yes it a better solution. Should be changed in Symfony 4 anyway.

Comment (theofidry) [inline src/Symfony/Component/Serializer/Normalizer/GetSetMethodNormalizer.php]: It would still be a problem:

```php
class MyCustomNormalizer extends AbstractObjectNormalizer
{
  protected instantiateObject(..., $allowedAttributes, $myParam = [])
  {
  }
```

And AFAIK it's a case where Symfony promises BC.

Comment (Ener-Getick) [inline src/Symfony/Component/Serializer/Normalizer/GetSetMethodNormalizer.php]: Take a look at my example: the new attribute is commented and fetched thanks to php functions so there is no bc problems.

Comment (dunglas) [inline src/Symfony/Component/Serializer/Normalizer/GetSetMethodNormalizer.php]: @theofidry can you implement @Ener-Getick solution?

Comment (theofidry) [inline src/Symfony/Component/Serializer/Normalizer/GetSetMethodNormalizer.php]: I'm on it, was just a bit too busy this week :)

Comment (theofidry) [inline src/Symfony/Component/Serializer/Normalizer/GetSetMethodNormalizer.php]: >Take a look at my example: the new attribute is commented and fetched thanks to php functions so there is no bc problems.

@Ener-Getick the problem I was showing was not a BC break introduced by the function signature, but by the fact that in this example `$myParam` cannot be taken as `$format`.

Comment (theofidry) [inline src/Symfony/Component/Serializer/Normalizer/GetSetMethodNormalizer.php]: @dunglas I see no easy fix for this one really, another solution is just accept that the current normalizers will now trigger a deprecation message error, which may be fixed manually by the users (if they care...) and will be fixed in the core in 4.0.

Comment (theofidry) [inline src/Symfony/Component/Serializer/Normalizer/GetSetMethodNormalizer.php]: An alternative approach to let the user change it easily, would be to create a protected function `getInstantiateObjectMethod(): string`, which defaults to `instantiateObject()` but can easily be overridden to return `instantiateComplexObject()` instead. This would allow the users to easily switch to the new method without too much work.

Comment (dunglas) [inline src/Symfony/Component/Serializer/Normalizer/GetSetMethodNormalizer.php]: @theofidry I've opened a PR on your fork implementing the solution proposed by @Ener-Getick: https://github.com/theofidry/symfony/pull/1

Comment (theofidry) [inline src/Symfony/Component/Serializer/Normalizer/GetSetMethodNormalizer.php]: @dunglas as said, my worries is that it's still a BC. If you have:

```php
class MyCustomNormalizer extends AbstractObjectNormalizer
{
  protected instantiateObject(..., $allowedAttributes, array $myParam = [])
  {
  }
}
```

Then you will have `$format = $myParam` which obviously will cause problems.

Comment (dunglas) [inline src/Symfony/Component/Serializer/Normalizer/GetSetMethodNormalizer.php]: When you override a protected method, you must respect its contract... It's not a BC break for me.

Comment (dunglas) [inline src/Symfony/Component/Serializer/Normalizer/GetSetMethodNormalizer.php]: And I doubt that anybody in the world has done something like that for this particular method.

Comment (dunglas) [inline src/Symfony/Component/Serializer/Normalizer/GetSetMethodNormalizer.php]: This is forbidden by the PHP interpreter itself: https://3v4l.org/S48gf
Si there is no problem here.

Comment (theofidry) [inline src/Symfony/Component/Serializer/Normalizer/GetSetMethodNormalizer.php]: >When you override a protected method, you must respect its contract... It's not a BC break for me.

If you override it this way there is not BC as the added param is optional.

In all honesty I would be happy to say yes to this solution it's the easiest way to deal with it really. The only thing I'm worried about is that it may introduce a BC break (but that's up to you to decide if it is or not) and the fact that the user may not see the change (i.e. see that he has to inject the format now).

Comment (dunglas) [inline src/Symfony/Component/Serializer/Normalizer/GetSetMethodNormalizer.php]: Anyway, it's not a BC break because the parent function is still called with the good number of arguments. See https://3v4l.org/EddB3.

Comment (theofidry) [inline src/Symfony/Component/Serializer/Normalizer/GetSetMethodNormalizer.php]: Nvm, even with my example it wouldn't be a problem as you would call the parent function with the right number of parameters. Not doing so is looking for trouble and would be the developer fault.

**Comment by dunglas** [comment]
Status: Reviewed

👍

File: src/Symfony/Component/Serializer/Normalizer/GetSetMethodNormalizer.php (hunk 13 above)
Comment (Ener-Getick) [inline src/Symfony/Component/Serializer/Normalizer/GetSetMethodNormalizer.php]: This solution is often used in symfony so i'd say it's not part of the symfony bc promises.

**Comment by dunglas** [comment]
Thank you @theofidry.

**Comment by theofidry** [comment]
You're welcome ;)
